codalab/server/bundle_manager.py

```diff
@@ -314,21 +314,45 @@ def _bring_offline_stuck_running_bundles(self, workers):
     def _schedule_run_bundles_on_workers(self, workers, staged_bundles_to_run, user_info_cache):
         """
         Schedule STAGED bundles to run on available workers based on the following logic:
-        1. If the bundle requests to run on a specific worker, tries to schedule the bundle
-           to run on a worker that has a tag exactly match with request_queue.
+        1. For a given user, schedule the highest-priority bundles first, followed by bundles
+           that request to run on a specific worker.
+        2. If the bundle requests to run on a specific worker, schedule the bundle
+           to run on a worker that has a tag that exactly matches the bundle's request_queue.
         2. If the bundle doesn't request to run on a specific worker,
           (1) try to schedule the bundle to run on a worker that belongs to the bundle's owner
           (2) if there is no such qualified private worker, uses CodaLab-owned workers, which have user ID root_user_id.
         :param workers: a WorkerInfoAccessor object containing worker related information e.g. running uuid.
         :param staged_bundles_to_run: a list of tuples each contains a valid bundle and its bundle resources.
         :param user_info_cache: a dictionary mapping user id to user information.
         """
-        # Reorder the stage_bundles so that bundles which were requested to run on a specific worker
-        # will be scheduled to run first
-        staged_bundles_to_run.sort(
-            key=lambda b: (b[0].metadata.request_queue is not None, b[0].metadata.request_queue),
-            reverse=True,
-        )
+        # Build a dictionary which maps from user id to positions in the queue of the
+        # user's staged bundles.
+        user_queue_positions = {}
+        for queue_position, staged_bundle in enumerate(staged_bundles_to_run):
+            if staged_bundle[0].owner_id not in user_queue_positions:
+                user_queue_positions[staged_bundle[0].owner_id] = []
+            user_queue_positions[staged_bundle[0].owner_id].append(queue_position)
+
+        for user, queue_positions in user_queue_positions.items():
+            sorted_queue_positions = sorted(queue_positions)
+            # Get this user's staged bundles
+            user_staged_bundles = [
+                staged_bundles_to_run[queue_position] for queue_position in queue_positions
+            ]
+            # Sort the staged bundles for this user, according to
+            # (1) their priority and (2) whether it requested to run on a specific worker.
```

Comment:
Make it clearer which is better (larger is higher priority, specific worker gets higher priority).